<h1>Содержание<span class="tocSkip"></span></h1>

<div class="toc">
    <ul class="toc-item">
        <li>
            <span>
                <a href="#Описание-проекта-—-Исследование-надёжности-заёмщиков" data-toc-modified-id="Описание-проекта-—-Исследование-надёжности-заёмщиков-1">
                    <span class="toc-item-num">
                        1&nbsp;&nbsp;
                    </span>
                    Описание проекта — Исследование надёжности заёмщиков
                </a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Цели-и-задачи-проекта" data-toc-modified-id="Цели-и-задачи-проекта-1.1">
                            <span class="toc-item-num">
                                1.1&nbsp;&nbsp;
                            </span>
                            Цели и задачи проекта
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.2">
                            <span class="toc-item-num">
                                1.2&nbsp;&nbsp;
                            </span>
                            Описание данных
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#План-работы" data-toc-modified-id="План-работы-1.3">
                            <span class="toc-item-num">
                                1.3&nbsp;&nbsp;
                            </span>
                            План работы
                        </a>
                    </span>
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Загрузка,-обзор-и-подготовка-данных" data-toc-modified-id="Загрузка,-обзор-и-подготовка-данных-2">
                    <span class="toc-item-num">
                        2&nbsp;&nbsp;
                    </span>
                    Загрузка, обзор и подготовка данных
                </a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2.1">
                            <span class="toc-item-num">
                                2.1&nbsp;&nbsp;
                            </span>
                            Загрузка данных
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2.2">
                            <span class="toc-item-num">
                                2.2&nbsp;&nbsp;
                            </span>
                            Обзор данных
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2.3">
                            <span class="toc-item-num">
                                2.3&nbsp;&nbsp;
                            </span>
                            Подготовка данных
                        </a>
                    </span>
                    <ul class="toc-item">
                        <li>
                            <span>
                                <a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2.3.1">
                                    <span class="toc-item-num">
                                        2.3.1&nbsp;&nbsp;
                                    </span>
                                    Заполнение пропусков
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Проверка-данных-на-аномалии" data-toc-modified-id="Проверка-данных-на-аномалии-2.3.2">
                                    <span class="toc-item-num">
                                        2.3.2&nbsp;&nbsp;
                                    </span>
                                    Проверка данных на аномалии
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.3.3">
                                    <span class="toc-item-num">
                                        2.3.3&nbsp;&nbsp;
                                    </span>
                                    Изменение типов данных
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-2.3.4">
                                    <span class="toc-item-num">
                                        2.3.4&nbsp;&nbsp;
                                    </span>
                                    Удаление дубликатов
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Создание-дополнительных-датафреймов" data-toc-modified-id="Создание-дополнительных-датафреймов-2.3.5">
                                    <span class="toc-item-num">
                                        2.3.5&nbsp;&nbsp;
                                    </span>
                                    Создание дополнительных датафреймов
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-2.3.6">
                                    <span class="toc-item-num">
                                        2.3.6&nbsp;&nbsp;
                                    </span>
                                    Категоризация дохода
                                </a>
                            </span>
                        </li>
                        <li>
                            <span>
                                <a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-2.3.7">
                                    <span class="toc-item-num">
                                        2.3.7&nbsp;&nbsp;
                                    </span>
                                    Категоризация целей кредита
                                </a>
                            </span>
                        </li>
                    </ul>    
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Анализ-зависимости-между-признаками" data-toc-modified-id="Анализ-зависимости-между-признаками-3">
                    <span class="toc-item-num">
                        3&nbsp;&nbsp;
                    </span>
                    Анализ зависимости между признаками
                </a>
            </span>
            <ul class="toc-item">
                <li>
                    <span>
                        <a href="#Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок-3.1">
                            <span class="toc-item-num">
                                3.1&nbsp;&nbsp;
                            </span>
                            Зависимость между количеством детей и возвратом кредита в срок
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок-3.2">
                            <span class="toc-item-num">
                                3.2&nbsp;&nbsp;
                            </span>
                            Зависимость между семейным положением и возвратом кредита в срок
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-3.3">
                            <span class="toc-item-num">
                                3.3&nbsp;&nbsp;
                            </span>
                            Зависимость между уровнем дохода и возвратом кредита в срок
                        </a>
                    </span>
                </li>
                <li>
                    <span>
                        <a href="#Зависимость-между-целью-кредита-и-возвратом-кредита-в-срок" data-toc-modified-id="Зависимость-между-целью-кредита-и-возвратом-кредита-в-срок-3.4">
                            <span class="toc-item-num">
                                3.4&nbsp;&nbsp;
                            </span>
                            Зависимость между целью кредита и возвратом кредита в срок
                        </a>
                    </span>
                </li>
            </ul>
        </li>
        <li>
            <span>
                <a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4">
                    <span class="toc-item-num">
                        4&nbsp;&nbsp;
                    </span>
                    Общий вывод
                </a>
            </span>
        </li>
    </ul>
</div>

# Описание проекта — Исследование надёжности заёмщиков
## Цели и задачи проекта

Заказчик — кредитный отдел банка. Нужно проанализировать, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

## Описание данных

Банк предоставил статистику о платёжеспособности клиентов в виде датасета со следующими столбцами:
+ `children` — количество детей в семье
+ `days_employed` — общий трудовой стаж в днях
+ `dob_years` — возраст клиента в годах
+ `education` — уровень образования клиента
+ `education_id` — идентификатор уровня образования
+ `family_status` — семейное положение
+ `family_status_id` — идентификатор семейного положения
+ `gender` — пол клиента
+ `income_type` — тип занятости
+ `debt` — имел ли задолженность по возврату кредитов
+ `total_income` — ежемесячный доход
+ `purpose` — цель получения кредита

## План работы

+ Загрузить и изучить данные
+ Провести предобработку данных
+ Проанализировать зависимость между возвратом кредита в срок и между:
    + количеством детей
    + семейным положением
    + уровнем дохода
    + целью кредита
+ Сделать общий вывод

# Загрузка, обзор и подготовка данных
## Загрузка данных

Импортируем библиотеку `pandas` и прочитаем файл `data.csv`

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data.csv')

## Обзор данных
Посмотрим первые и последние 10 строк таблицы `df`

In [3]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
df.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Посмотрим общую информацию о данных в таблице `df`

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Посмотрим названия колонок таблицы `df`

In [6]:
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

ИТОГО:
1. Мы работаем с таблицей размером 21525 строк на 12 столбцов.
2. Проблем в названии колонок нет.
3. Количество значений в столбцах различается. Значит в данных есть пропущенные значения.
4. Общий трудовой стаж в днях представлен как `float`, но по своей сути это `int`.
5. Общий трудовой стаж является отрицательным значением в некоторой части строк.
6. Общий трудовой стаж в некоторых ячейках превышает 900 лет.
7. В столбце `education` есть дубликаты написанные в разном регистре.

## Подготовка данных
### Заполнение пропусков

Посчитаем количество пропущенных значений в таблице `df`

In [7]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Количество пропусков в столбцах `days_employed` и `total_income` совпадает. Скорее всего пропуски сделаны в одинаковых строках. Создадим таблицу `skip_df`, чтобы проверить это. В неё войдут только те строки, в которых есть пропуски как в столбце `days_employed`, так и в столбце `total_income`.

In [8]:
skip_df = df.loc[df['total_income'].isna() & df['days_employed'].isna()]
len(skip_df)

2174

Количество строк таблицы `skip_df` совпадает с количеством пропусков, полученных ранее. Значит пропуски столбцов `days_employed` и `total_income` сделаны в одинаковых строках. Посмотрим первые и последние 10 строк таблицы `skip_df`.

In [9]:
skip_df.head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [10]:
skip_df.tail(10)

children  days_employed  dob_years            education  education_id  \
21415         0            NaN         54              среднее             1   
21423         0            NaN         63              среднее             1   
21426         0            NaN         49              среднее             1   
21432         1            NaN         38  неоконченное высшее             2   
21463         1            NaN         35               высшее             0   
21489         2            NaN         47              Среднее             1   
21495         1            NaN         50              среднее             1   
21497         0            NaN         48               ВЫСШЕЕ             0   
21502         1            NaN         42              среднее             1   
21510         2            NaN         28              среднее             1   

               family_status  family_status_id gender income_type  debt  \
21415        женат / замужем                 0      F   пенсионер     0   
21423        женат / замужем                 0      M   пенсионер     0   
21426        женат / замужем                 0      F   сотрудник     1   
21432  Не женат / не замужем                 4      F   сотрудник     0   
21463       гражданский брак                 1      M   сотрудник     0   
21489        женат / замужем                 0      M   компаньон     0   
21495       гражданский брак                 1      F   сотрудник     0   
21497        женат / замужем                 0      F   компаньон     0   
21502        женат / замужем                 0      F   сотрудник     0   
21510        женат / замужем                 0      F   сотрудник     0   

       total_income                           purpose  
21415           NaN                 операции с жильем  
21423           NaN              сделка с автомобилем  
21426           NaN                      недвижимость  
21432           NaN                 операции с жильем  
21463           NaN             на проведение свадьбы  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля

ИТОГО:
1. Пропущенные значения в столбцах `days_employed` и `total_income` заполнены значениями NaN.
2. Доля пропущенных данных составляет 10% от общего объёма.
3. Так как пропуски столбцов `days_employed` и `total_income` не хаотичны, а сделаны в одинаковых строках, значит они привязаны к конкретным людям. А это в свою очередь может значить, что причиной появления пропусков мог быть либо человеческий фактор (данные просто не внесли в полном объёме), либо эти 2174 человека не предоставили данные трудового стажа и ежемесячного дохода.
4. Простое удаление строк с пустыми ячейками приведёт к потере 10% записей о семейном положении и количестве детей, поэтому лучше заменить их каким-нибудь другим значением. Так как значения в столбцах `days_employed` и `total_income` являются количественными, то в них могут быть значительные отклонения в сторону минимума или максимума, поэтому лучшим решением будет заполнить пропуски медианным значением. Но сперва следует избавиться от аномалий, которые присутствую в этих столбцах.

### Проверка данных на аномалии

Разберёмся с аномалиями в столбце `days_employed`. Здесь большая часть данных является отрицательным значением, а все положительные значения сверхбольшие. Сперва найдём минимум и максимум среди всех положительных значений трудового стажа, чтобы понять размер диапазона.

In [11]:
row = df['days_employed']>=0
df['days_employed'].loc[row].min()

328728.72060451825

In [12]:
df['days_employed'].loc[row].max()

401755.40047533

Получается что все положительные значения трудового стажа лежат в диапазоне от 900 до 1100 лет. Следует проверить возраст и количество людей с таким аномальным стажем.

In [13]:
# Количество людей с аномальным стажем
len(df['days_employed'].loc[row])

3445

In [14]:
# Минимальный возраст среди людей с аномальным стажем
df['dob_years'].loc[row].min()

0

In [15]:
# Максимальный возраст среди людей с аномальным стажем
df['dob_years'].loc[row].max()

74

Получается, что это 3445 человек в возрасте от 0 до 74 лет и они составляют 15.9% от общего количества данных в нашей таблице, поэтому мы не можем просто удалить эти строки. Посмотрим сколько среди них возрастных людей старше 50 лет.

In [16]:
people_over_50 = (df['dob_years']>50) & (df['days_employed']>=0)
len(df.loc[people_over_50])

3202

Получается, что 93% людей с аномальным стажем старше 50 лет. Если разделить весь положительный трудовой стаж на 20, тогда он будет находиться в диапазоне от 45 до 55 лет, что вполне нормально для людей пенсионного возраста.

Теперь рассмотрим отрицательные значения трудового стажа. Так же как и для положительных значений, найдём минимум и максимум среди всех отрицательных значений трудового стажа.

In [17]:
df['days_employed'].loc[df['days_employed']<0].min()

-18388.949900568383

In [18]:
df['days_employed'].loc[df['days_employed']<0].max()

-24.14163324048118

Если не обращать внимания на знак минус, то все отрицательные значения трудового стажа лежат в диапазоне от 24 дней до 50 лет, что вполне нормально, поэтому просто избавимся от знака минус взяв все отрицательные значения трудового стажа по модулю.

In [19]:
# Разделим все положительные значения трудового стажа на 20
df.loc[df['days_employed']>=0, 'days_employed'] = df.loc[df['days_employed']>=0, 'days_employed']/20
# Возьмём все отрицательные значения трудового стажа по модулю
df.loc[df['days_employed']<0, 'days_employed'] = df.loc[df['days_employed']<0, 'days_employed'].abs()
# Посмотрим первые 10 строк таблицы df после сделанных изменений
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   17013.303602         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Теперь можно заполнить пустые ячейки столбца `days_employed` медианным значением.

In [20]:
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())

Также проверим на аномалии столбцы `dob_years`, `education_id`, `family_status_id` и `debt`. Посмотрим их уникальные значения.

In [21]:
df['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75], dtype=int64)

In [22]:
df['education_id'].sort_values().unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [23]:
df['family_status_id'].sort_values().unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [24]:
df['debt'].sort_values().unique()

array([0, 1], dtype=int64)

Наблюдается одна единственная аномалия. Это люди с возрастом 0 лет. Посмотрим сколько таких людей и кто они.

In [25]:
len(df.loc[df['dob_years']==0])

101

In [26]:
df.loc[df['dob_years']==0].head()

children  days_employed  dob_years education  education_id  \
99           0   17327.080945          0   Среднее             1   
149          0    2664.273168          0   среднее             1   
270          3    1872.663186          0   среднее             1   
578          0   19892.828251          0   среднее             1   
1040         0    1158.029561          0    высшее             0   

        family_status  family_status_id gender income_type  debt  \
99    женат / замужем                 0      F   пенсионер     0   
149         в разводе                 3      F   сотрудник     0   
270   женат / замужем                 0      F   сотрудник     0   
578   женат / замужем                 0      F   пенсионер     0   
1040        в разводе                 3      F   компаньон     0   

       total_income                                 purpose  
99     71291.522491                              автомобиль  
149    70176.435951                       операции с жильем  
270   102166.458894                            ремонт жилью  
578    97620.687042  строительство собственной недвижимости  
1040  303994.134987                         свой автомобиль

In [27]:
df.loc[df['dob_years']==0].tail()

children  days_employed  dob_years education  education_id  \
19829         0    2194.220567          0   среднее             1   
20462         0   16936.743427          0   среднее             1   
20577         0   16587.063573          0   среднее             1   
21179         2     108.967042          0    высшее             0   
21313         0    1268.487728          0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
19829        женат / замужем                 0      F   сотрудник     0   
20462        женат / замужем                 0      F   пенсионер     0   
20577  Не женат / не замужем                 4      F   пенсионер     0   
21179        женат / замужем                 0      M   компаньон     0   
21313  Не женат / не замужем                 4      M   сотрудник     0   

        total_income                           purpose  
19829            NaN                             жилье  
20462  259193.920299              покупка своего жилья  
20577  129788.762899                      недвижимость  
21179  240702.007382  строительство жилой недвижимости  
21313  152419.184851                 операции с жильем

Таких людей всего 101 или 0,47% от общего объёма данных. Судя по столбцам `children`, `days_employed` и `income_type` - это люди самых разных возрастных категорий, поэтому заменить все нули на какое-либо одно усреднённое значение возраста не выйдет. Оставим как есть. На главную цель исследования это никак не повлияет.

Осталось проверить столбец `total_income`. Здесь просмотр уникальных значений ничего не даст. Пожалуй единственная аномалия, которая может быть в этом столбце - это отрицательный доход. Поэтому просто проверим столбец `total_income` на наличие отрицательных значений.

In [28]:
len(df.loc[df['total_income']<0])

0

Таких значений не оказалось, значит всё в порядке.

### Изменение типов данных

Значения в столбцах `days_employed` и `total_income` имеют тип `float`, но все цифры после запятой ни на что не влияют и усложняют визуальное восприятие данных. Преобразуем их в целочисленный тип `int`. Так как в столбце `total_income` есть пропуски, а преобразование `NaN` в `int` может вызвать ошибку, то используем конструкцию `try-except`. Также заполним пропуски в столбце `total_income` медианным значением по аналогии с заполнением в столбце `days_employed`.

In [29]:
try:
    df['total_income'] = df['total_income'].astype('int')
    df['days_employed'] = df['days_employed'].astype('int')
    df['total_income'] = df['total_income'].fillna(df['total_income'].median())
except:
    df['total_income'] = df['total_income'].fillna(df['total_income'].median())
    df['total_income'] = df['total_income'].astype('int')
    df['days_employed'] = df['days_employed'].astype('int')

Посмотрим общую информацию о данных в таблице `df` после сделанных изменений.

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


### Удаление дубликатов
Займёмся очисткой данных от дубликатов. Посчитаем и удалим явные дубликаты таблицы `df`.

In [31]:
df.duplicated().sum()

54

In [32]:
df = df.drop_duplicates().reset_index(drop=True)

Проверим наличие явных дубликатов и посмотрим общую информацию о данных в таблице `df` после сделанных изменений.

In [33]:
df.duplicated().sum()

0

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21471 non-null  int64 
 1   days_employed     21471 non-null  int32 
 2   dob_years         21471 non-null  int64 
 3   education         21471 non-null  object
 4   education_id      21471 non-null  int64 
 5   family_status     21471 non-null  object
 6   family_status_id  21471 non-null  int64 
 7   gender            21471 non-null  object
 8   income_type       21471 non-null  object
 9   debt              21471 non-null  int64 
 10  total_income      21471 non-null  int32 
 11  purpose           21471 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


Теперь разберёмся с неявными дубликатами. Посмотрим уникальные значения всех столбцов имеющих тип данных `object`.

In [35]:
df['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

In [36]:
df['family_status'].sort_values().unique()

array(['Не женат / не замужем', 'в разводе', 'вдовец / вдова',
       'гражданский брак', 'женат / замужем'], dtype=object)

In [37]:
df['gender'].sort_values().unique()

array(['F', 'M', 'XNA'], dtype=object)

In [38]:
df['income_type'].sort_values().unique()

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

In [39]:
df['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

В столбцах `education` и `family_status` значения записаны в разном регистре. Приведём все значения в этих столбцах к нижнему регистру.

In [40]:
df['education'] = df['education'].str.lower()
df['family_status'] = df['family_status'].str.lower()

Проверим уникальные значения в столбцах `education` и `family_status` после сделанных изменений.

In [41]:
df['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

In [42]:
df['family_status'].sort_values().unique()

array(['в разводе', 'вдовец / вдова', 'гражданский брак',
       'женат / замужем', 'не женат / не замужем'], dtype=object)

### Создание дополнительных датафреймов

Создадим два новых датафрейма `education_df` и `family_df`, в которых каждому уникальному значению из `education` и `family_status` таблицы `df` соответствует уникальное значение `education_id` и `family_status_id`.

In [43]:
education_data = [
    ['начальное', 3],
    ['среднее', 1],
    ['неоконченное высшее', 2],
    ['высшее', 0],
    ['ученая степень', 4]
]

family_data = [
    ['женат / замужем', 0],
    ['не женат / не замужем', 4],
    ['гражданский брак', 1],
    ['в разводе', 3],
    ['вдовец / вдова', 2]
]
education_columns = ['education', 'education_id']
family_columns = ['family_status', 'family_status_id']

education_df = pd.DataFrame(data=education_data, columns=education_columns)
family_df = pd.DataFrame(data=family_data, columns=family_columns)

In [44]:
education_df

education  education_id
0            начальное             3
1              среднее             1
2  неоконченное высшее             2
3               высшее             0
4       ученая степень             4

In [45]:
family_df

family_status  family_status_id
0        женат / замужем                 0
1  не женат / не замужем                 4
2       гражданский брак                 1
3              в разводе                 3
4         вдовец / вдова                 2

Удалим из исходного датафрейма `df` столбцы `education` и `family_status`. Посмотрим первые 10 строк после сделанных изменений.

In [46]:
df = df.drop(columns=['education', 'family_status'])
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу  
5   компаньон     0        255763               покупка жилья  
6   компаньон     0        240525           операции с жильем  
7   сотрудник     0        135823                 образование  
8   сотрудник     0         95856       на проведение свадьбы  
9   сотрудник     0        144425     покупка жилья для семьи

### Категоризация дохода

Добавим в нашу таблицу `df` столбец `total_income_category`, в котором будем присваивать клиентам банка категории, в зависимости от уровня ежемесячного дохода. Посмотрим первые 10 строк таблицы `df` после сделанных изменений.

In [47]:
def change_total_income(row):
    line = row['total_income']
    if 0<=line<=30000:
        return 'E'
    elif 30001<=line<=50000:
        return 'D'
    elif 50001<=line<=200000:
        return 'C'
    elif 200001<=line<=1000000:
        return 'B'
    elif line>=1000001:
        return 'A'
    
df['total_income_category'] = df.apply(change_total_income, axis=1)
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

### Категоризация целей кредита

Напишем функцию `change_purpose_credit` и используя метод `apply` добавим в нашу таблицу `df` новый столбец `purpose_category`, в котором мы обобщим цели получения кредита на основе столбца `purpose`. Посмотрим первые 10 строк таблицы `df` после сделанных изменений.

In [48]:
def change_purpose_credit(row):
    line = row['purpose']
    if 'автомобил' in line:
        return 'операции с автомобилем'
    elif 'жиль' in line or 'недвижимост' in line:
        return 'операции с недвижимостью'
    elif 'свадьб' in line:
        return 'проведение свадьбы'
    elif 'образован' in line:
        return 'получение образования'
    else:
        return 'нет категории'

df['purpose_category'] = df.apply(change_purpose_credit, axis=1)
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

# Анализ зависимости между признаками
Чтобы проанализировать зависимости, сделаем несколько сводных таблиц. Мы будем расчитывать среднеарифметическое значение столбца `debt` для каждого уникального значения из столбцов `children`, `family_status_id`, `total_income_category` и `purpose_category`. В столбце `debt` значение 0 - это отсутствие задолженности в прошлом, а значение 1 - это наличие задолженности в прошлом, поэтому чем ближе среднеарифметическое будет к 0, тем более надёжным будет клиент.

## Зависимость между количеством детей и возвратом кредита в срок

In [49]:
df.pivot_table(index='children', values='debt', aggfunc=['mean', 'count'])

mean  count
              debt   debt
children                 
-1        0.021277     47
 0        0.075353  14107
 1        0.092327   4809
 2        0.094542   2052
 3        0.081818    330
 4        0.097561     41
 5        0.000000      9
 20       0.105263     76

Обнаружилась аномалия, которую не учли на стадии предобработки данных. Значения с количеством детей -1 и 20 (-1 невозможно, а 20 крайне маловероятно). Среднеарифметическое по столбцу `debt` для 20 детей очень близко к значениям для 1, 2, 3 или 4 детей. Примем что это многодетные семьи с количеством детей более 5. Ячейки со значением -1 столбца `children` заменим на значение 0. После сделанных изменений заново построим сводную таблицу.

In [50]:
df.loc[df['children']==-1, 'children'] = 0
df.pivot_table(index='children', values='debt', aggfunc=['mean', 'count'])

mean  count
              debt   debt
children                 
0         0.075173  14154
1         0.092327   4809
2         0.094542   2052
3         0.081818    330
4         0.097561     41
5         0.000000      9
20        0.105263     76

Также стоит обратить внимание, что при количестве детей 5 среднеарифметическое значение `debt` равно ровно нулю. Очевидно, что это связано с тем, что людей с таким количеством детей всего 9 (или 0,04% от общего количества клиентов) и ни у кого из них не было задолженностей.

__Вывод:__ Количество детей __влияет__ на вероятность возврата кредита в срок. Чем меньше детей, тем больше вероятность возврата кредита в срок.  Из общей тенденции выделяются только люди с 3 детьми, но это скорее результат малости выборки (всего 330 клиентов банка из 21471), чем исключение.

## Зависимость между семейным положением и возвратом кредита в срок

In [51]:
df.pivot_table(index=['family_status_id'], values=['debt'], aggfunc=['mean', 'count'])

mean  count
                      debt   debt
family_status_id                 
0                 0.075421  12344
1                 0.093202   4163
2                 0.065693    959
3                 0.071130   1195
4                 0.097509   2810

Для полноты картины посмотрим таблицу `family_df`.

In [52]:
family_df

family_status  family_status_id
0        женат / замужем                 0
1  не женат / не замужем                 4
2       гражданский брак                 1
3              в разводе                 3
4         вдовец / вдова                 2

__Вывод:__ Семейное положение __влияет__ на вероятность возврата кредита в срок. Наиболее надёжные клиенты - это люди состоящие в гражданском браке.

## Зависимость между уровнем дохода и возвратом кредита в срок

In [53]:
df.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['mean', 'count'])

mean  count
                           debt   debt
total_income_category                 
A                      0.080000     25
B                      0.070621   5041
C                      0.084825  16033
D                      0.060000    350
E                      0.090909     22

__Вывод:__ Логично предположить, что чем выше уровень дохода, тем более надёжный должен быть клиент, но в сводной таблице не наблюдается такой закономерности, поэтому можно сказать, что уровень дохода __не влияет__ на вероятность возврата кредита в срок.

## Зависимость между целью кредита и возвратом кредита в срок

In [54]:
df.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['mean', 'count'])

mean  count
                              debt   debt
purpose_category                         
операции с автомобилем    0.093547   4308
операции с недвижимостью  0.072314  10814
получение образования     0.092177   4014
проведение свадьбы        0.079657   2335

__Вывод:__ Цель кредита __влияет__ на вероятность возврата кредита в срок. Операции с автомобилем и получение образования менее надёжны с точки зрения возврата кредита в срок, чем операции с недвижимостью или проведение свадьбы.

# Общий вывод

Количество должников, относительно общего количества клиентов, находится в диапазоне от 6,5% до 10,5%. Семейное положение и количество детей клиента __влияет__ на вероятность погашения кредита в срок, но значимость этого влияния находится в диапазоне от 0% до 3%.